File: src/main/java/net/gtaun/shoebill/amx/AmxInstanceImpl.java
package net.gtaun.shoebill.amx;

import net.gtaun.shoebill.SampNativeFunction;

import java.util.HashMap;
import java.util.function.Function;

public class AmxInstanceImpl implements AmxInstance {

    private final int handle;
    private HashMap<String, Function<Object[], Integer>> registeredFunctions;

    public AmxInstanceImpl(int handle) {
        this.handle = handle;
        this.registeredFunctions = new HashMap<>();
    }

    public int getHandle() {
        return handle;
    }

    @Override
    public AmxCallable getPublic(String name) {
        int funcHandle = SampNativeFunction.getPublic(handle, name);
        if (funcHandle == AmxCallable.INVALID_CALLABLE)
            return null;
        return new AmxCallableImpl(this, funcHandle, name, AmxCallableType.PUBLIC);
    }

    @Override
    public AmxCallable getNative(String name) {
        int funcHandle = SampNativeFunction.getNative(handle, name);
        if (funcHandle == AmxCallable.INVALID_CALLABLE)
            return null;
        return new AmxCallableImpl(this, funcHandle, name, AmxCallableType.NATIVE);
    }

    @Override
    public boolean registerFunction(String s, Function<Object[], Integer> function, Class... classes) {
        if(registeredFunctions.containsKey(s))
            return false;
        boolean result = SampNativeFunction.registerFunction(handle, s, classes);
        if(result) registeredFunctions.put(s, function);
        return result;
    }

    @Override
    public boolean unregisterFunction(String s) {
        if(!registeredFunctions.containsKey(s))
            return false;
        boolean result = SampNativeFunction.unregisterFunction(handle, s);
        if(result) registeredFunctions.remove(s);
        return result;
    }

    @Override
    public boolean hasRegisteredFunction(String name) {
        return registeredFunctions.containsKey(name);
    }

    @Override
    public int callRegisteredFunction(String name, Object... parameters) {
        if(!registeredFunctions.containsKey(name))
            return -1;
        return registeredFunctions.get(name).apply(parameters);
    }


}

File: src/main/java/net/gtaun/shoebill/amx/AmxInstanceImpl.kt
package net.gtaun.shoebill.amx

import net.gtaun.shoebill.SampNativeFunction
import net.gtaun.shoebill.amx.types.ReturnType
import java.util.function.Function

/**
 * @author Marvin Haschker
 */
internal class AmxInstanceImpl(override val handle: Int) : AmxInstance() {

    private val registeredFunctions: MutableMap<String, (Array<Any>) -> Int> = mutableMapOf()

    override fun getPublic(name: String, returnType: ReturnType): AmxCallable? {
        val funcHandle = SampNativeFunction.getPublic(handle, name)
        if (funcHandle == AmxCallable.INVALID_CALLABLE)
            return null
        return AmxCallableImpl(this, funcHandle, name, AmxCallableType.PUBLIC, returnType)
    }

    override fun getNative(name: String, returnType: ReturnType): AmxCallable? {
        val id = SampNativeFunction.getNative(name)
        return if (id != 0)
            AmxCallableImpl(this, id, name, AmxCallableType.NATIVE, returnType)
        else
            null
    }

    override fun unregisterFunction(name: String): Boolean {
        if (!registeredFunctions.containsKey(name))
            return false
        val result = SampNativeFunction.unregisterFunction(handle, name)
        if (result) registeredFunctions.remove(name)
        return result
    }

    override fun callRegisteredFunction(name: String, parameters: Array<Any>): Int {
        return registeredFunctions[name]?.invoke(parameters) ?: -1
    }

    override fun isFunctionRegistered(name: String): Boolean = registeredFunctions.containsKey(name)

    override fun registerFunction(name: String, callback: (Array<Any>) -> Int,
                                  vararg types: String): Boolean {
        if (registeredFunctions.containsKey(name)) return false
        val result = SampNativeFunction.registerFunction(handle, name, *types)
        if (result) registeredFunctions[name] = callback
        return result
    }
}
